File: backend-spring/src/main/java/com/ivyts/backend/service/AuthService.java
package com.ivyts.backend.service;

import com.ivyts.backend.common.exception.ApiException;
import com.ivyts.backend.domain.user.PendingEmailChange;
import com.ivyts.backend.domain.user.PendingPhoneChange;
import com.ivyts.backend.domain.user.User;
import com.ivyts.backend.domain.user.UserRole;
import com.ivyts.backend.notification.NotificationEventsService;
import com.ivyts.backend.security.AuthUser;
import com.ivyts.backend.security.JwtService;
import com.ivyts.backend.service.userstore.UserStore;
import com.ivyts.backend.web.auth.dto.AuthResponse;
import com.ivyts.backend.web.auth.dto.ChangePasswordRequest;
import com.ivyts.backend.web.auth.dto.ConfirmEmailChangeRequest;
import com.ivyts.backend.web.auth.dto.ConfirmPhoneChangeRequest;
import com.ivyts.backend.web.auth.dto.LoginRequest;
import com.ivyts.backend.web.auth.dto.PublicUserResponse;
import com.ivyts.backend.web.auth.dto.RefreshTokenRequest;
import com.ivyts.backend.web.auth.dto.RegisterRequest;
import com.ivyts.backend.web.auth.dto.RequestEmailChangeRequest;
import com.ivyts.backend.web.auth.dto.RequestPhoneChangeRequest;
import com.ivyts.backend.web.auth.dto.UpdateProfileRequest;
import com.ivyts.backend.web.auth.dto.VerificationResponse;
import java.time.Instant;
import java.util.List;
import java.util.Locale;
import java.util.Random;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

@Service
public class AuthService {

    private static final BCryptPasswordEncoder PASSWORD_ENCODER = new BCryptPasswordEncoder(10);
    private final UserStore userStore;
    private final JwtService jwtService;
    private final NotificationEventsService notificationEventsService;
    private final Random random = new Random();

    public AuthService(UserStore userStore, JwtService jwtService, NotificationEventsService notificationEventsService) {
        this.userStore = userStore;
        this.jwtService = jwtService;
        this.notificationEventsService = notificationEventsService;
    }

    public AuthResponse register(RegisterRequest request) {
        String normalizedEmail = normalizeEmail(request.email());
        if (userStore.findByEmail(normalizedEmail).isPresent()) {
            throw new ApiException(HttpStatus.CONFLICT, "Email already exists");
        }

        User user = new User();
        user.setFullName(request.fullName().trim());
        user.setEmail(normalizedEmail);
        user.setPasswordHash(PASSWORD_ENCODER.encode(request.password()));
        user.setPhone(blankToNull(request.phone()));
        user.setRole(UserRole.STUDENT);
        user.setActive(true);
        user = userStore.save(user);
        notificationEventsService.emitNewUserRegistered(user.getId(), user.getEmail(), user.getFullName());

        return buildAuthResponse(user);
    }

    public AuthResponse login(LoginRequest request) {
        User user = userStore.findByEmail(normalizeEmail(request.email()))
            .orElseThrow(() -> new ApiException(HttpStatus.UNAUTHORIZED, "Email or password is incorrect"));

        ensureActive(user);

        if (!PASSWORD_ENCODER.matches(request.password(), user.getPasswordHash())) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, "Email or password is incorrect");
        }

        return buildAuthResponse(user);
    }

    public void logout(AuthUser authUser) {
        User user = findUserByIdOrThrow(authUser.userId());
        user.setRefreshToken(null);
        userStore.save(user);
    }

    public PublicUserResponse me(AuthUser authUser) {
        return toPublicUser(findUserByIdOrThrow(authUser.userId()));
    }

    public AuthResponse refreshToken(RefreshTokenRequest request) {
        AuthUser payload;
        try {
            payload = jwtService.verifyRefreshToken(request.refreshToken());
        } catch (Exception exception) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, "Invalid or expired refresh token");
        }

        User user = findUserByIdOrThrow(payload.userId());
        ensureActive(user);

        if (user.getRefreshToken() == null || !user.getRefreshToken().equals(request.refreshToken())) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, "Refresh token is not valid");
        }

        return buildAuthResponse(user);
    }

    public PublicUserResponse updateProfile(AuthUser authUser, UpdateProfileRequest request) {
        User user = findUserByIdOrThrow(authUser.userId());
        ensureActive(user);

        if (request.fullName() != null) {
            user.setFullName(request.fullName().trim());
        }
        if (request.avatarUrl() != null) {
            user.setAvatarUrl(blankToNull(request.avatarUrl()));
        }
        if (request.bio() != null) {
            user.setBio(blankToNull(request.bio()));
        }

        userStore.save(user);
        return toPublicUser(user);
    }

    public void changePassword(AuthUser authUser, ChangePasswordRequest request) {
        User user = findUserByIdOrThrow(authUser.userId());
        ensureActive(user);

        if (!PASSWORD_ENCODER.matches(request.currentPassword(), user.getPasswordHash())) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "Current password is incorrect");
        }
        if (PASSWORD_ENCODER.matches(request.newPassword(), user.getPasswordHash())) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "New password must be different from current password");
        }

        user.setPasswordHash(PASSWORD_ENCODER.encode(request.newPassword()));
        userStore.save(user);
    }

    public VerificationResponse requestEmailChange(AuthUser authUser, RequestEmailChangeRequest request) {
        User user = findUserByIdOrThrow(authUser.userId());
        ensureActive(user);

        String newEmail = normalizeEmail(request.newEmail());
        if (newEmail.equals(user.getEmail())) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "New email must be different from current email");
        }
        userStore.findByEmail(newEmail)
            .filter(existing -> !existing.getId().equals(user.getId()))
            .ifPresent(existing -> { throw new ApiException(HttpStatus.CONFLICT, "Email already exists"); });

        String code = generateSixDigitCode();
        Instant expiresAt = Instant.now().plusSeconds(10 * 60);
        user.setPendingEmailChange(new PendingEmailChange(newEmail, code, expiresAt));
        userStore.save(user);

        return new VerificationResponse(newEmail, expiresAt, code);
    }

    public AuthResponse confirmEmailChange(AuthUser authUser, ConfirmEmailChangeRequest request) {
        User user = findUserByIdOrThrow(authUser.userId());
        ensureActive(user);

        PendingEmailChange pending = user.getPendingEmailChange();
        if (pending == null) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "No pending email verification request");
        }

        String newEmail = normalizeEmail(request.newEmail());
        if (!pending.newEmail().equals(newEmail)) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "Pending email does not match");
        }
        if (pending.expiresAt().isBefore(Instant.now())) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "Email verification code has expired");
        }
        if (!pending.verificationCode().equals(request.verificationCode())) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "Email verification code is incorrect");
        }

        userStore.findByEmail(newEmail)
            .filter(existing -> !existing.getId().equals(user.getId()))
            .ifPresent(existing -> { throw new ApiException(HttpStatus.CONFLICT, "Email already exists"); });

        user.setEmail(newEmail);
        user.setPendingEmailChange(null);
        userStore.save(user);
        return buildAuthResponse(user);
    }

    public VerificationResponse requestPhoneChange(AuthUser authUser, RequestPhoneChangeRequest request) {
        User user = findUserByIdOrThrow(authUser.userId());
        ensureActive(user);

        String newPhone = request.newPhone().trim();
        if (newPhone.equals(user.getPhone())) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "New phone must be different from current phone");
        }

        String code = generateSixDigitCode();
        Instant expiresAt = Instant.now().plusSeconds(10 * 60);
        user.setPendingPhoneChange(new PendingPhoneChange(newPhone, code, expiresAt));
        userStore.save(user);

        return new VerificationResponse(newPhone, expiresAt, code);
    }

    public PublicUserResponse confirmPhoneChange(AuthUser authUser, ConfirmPhoneChangeRequest request) {
        User user = findUserByIdOrThrow(authUser.userId());
        ensureActive(user);

        PendingPhoneChange pending = user.getPendingPhoneChange();
        if (pending == null) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "No pending phone verification request");
        }
        if (!pending.newPhone().equals(request.newPhone().trim())) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "Pending phone does not match");
        }
        if (pending.expiresAt().isBefore(Instant.now())) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "Phone OTP has expired");
        }
        if (!pending.otpCode().equals(request.otpCode())) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "Phone OTP is incorrect");
        }

        user.setPhone(request.newPhone().trim());
        user.setPendingPhoneChange(null);
        userStore.save(user);
        return toPublicUser(user);
    }

    private AuthResponse buildAuthResponse(User user) {
        AuthUser payload = new AuthUser(user.getId(), user.getEmail(), user.getRole());
        String accessToken = jwtService.signAccessToken(payload);
        String refreshToken = jwtService.signRefreshToken(payload);
        user.setRefreshToken(refreshToken);
        User persistedUser = userStore.save(user);
        return new AuthResponse(toPublicUser(persistedUser), accessToken, refreshToken);
    }

    private PublicUserResponse toPublicUser(User user) {
        return new PublicUserResponse(
            user.getId(),
            user.getFullName(),
            user.getEmail(),
            user.getRole().name().toLowerCase(Locale.ROOT),
            user.getAvatarUrl(),
            user.getPhone(),
            user.getBio(),
            user.isActive(),
            user.getOwnedCourseIds() == null ? List.of() : user.getOwnedCourseIds()
        );
    }

    private User findUserByIdOrThrow(String userId) {
        return userStore.findById(userId)
            .orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "User not found"));
    }

    private void ensureActive(User user) {
        if (!user.isActive()) {
            throw new ApiException(HttpStatus.FORBIDDEN, "Account is inactive");
        }
    }

    private String generateSixDigitCode() {
        return String.format(Locale.ROOT, "%06d", random.nextInt(1_000_000));
    }

    private String normalizeEmail(String email) {
        return email.trim().toLowerCase(Locale.ROOT);
    }

    private String blankToNull(String value) {
        if (value == null) {
            return null;
        }
        String trimmed = value.trim();
        return trimmed.isEmpty() ? null : trimmed;
    }
}

File: archive/backend-node/src/services/auth.service.ts
// Authentication, token rotation and current-user self-service profile flows.
import bcrypt from 'bcrypt';
import { User, UserHydratedDocument } from '../models/user.model.js';
import { HTTP_STATUS } from '../constants/http-status.js';
import { ApiError } from '../utils/api-error.js';
import { mapUserToAuthPayload, sanitizeUser } from '../utils/auth-user.js';
import {
  ChangePasswordInput,
  ConfirmEmailChangeInput,
  ConfirmPhoneChangeInput,
  LoginInput,
  RefreshTokenInput,
  RegisterInput,
  RequestEmailChangeInput,
  RequestPhoneChangeInput,
  UpdateProfileInput,
} from '../validations/auth.validation.js';
import { signAccessToken, signRefreshToken, verifyRefreshToken } from '../utils/jwt.js';
import { emitNewUserRegistered, emitProfileUpdated } from './notification-events.service.js';

const SALT_ROUNDS = 10;

interface AuthResponseData {
  user: ReturnType<typeof sanitizeUser>;
  accessToken: string;
  refreshToken: string;
}

interface VerificationRequestResult {
  deliveryTarget: string;
  expiresAt: Date;
  verificationPreviewCode: string;
}

async function buildAuthResponse(userId: string): Promise<AuthResponseData> {
  const user = await User.findById(userId);

  if (!user) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'User not found');
  }

  if (!user.isActive) {
    throw new ApiError(HTTP_STATUS.FORBIDDEN, 'Account is inactive');
  }

  const payload = mapUserToAuthPayload(user);
  const accessToken = signAccessToken(payload);
  const refreshToken = signRefreshToken(payload);

  await User.findByIdAndUpdate(user._id, { refreshToken });

  return {
    user: sanitizeUser(user),
    accessToken,
    refreshToken,
  };
}

async function findUserByIdOrThrow(userId: string): Promise<UserHydratedDocument> {
  const user = await User.findById(userId).select('+passwordHash');

  if (!user) {
    throw new ApiError(HTTP_STATUS.NOT_FOUND, 'User not found');
  }

  if (!user.isActive) {
    throw new ApiError(HTTP_STATUS.FORBIDDEN, 'Account is inactive');
  }

  return user;
}

function generateSixDigitCode(): string {
  return String(Math.floor(100000 + Math.random() * 900000));
}

export async function registerUser(input: RegisterInput): Promise<AuthResponseData> {
  const existingUser = await User.findOne({ email: input.email });

  if (existingUser) {
    throw new ApiError(HTTP_STATUS.CONFLICT, 'Email already exists');
  }

  const passwordHash = await bcrypt.hash(input.password, SALT_ROUNDS);

  const user = await User.create({
    fullName: input.fullName,
    email: input.email,
    passwordHash,
    phone: input.phone,
    role: 'student',
  });

  emitNewUserRegistered(user._id.toString(), user.email, user.fullName);

  return buildAuthResponse(user._id.toString());
}

export async function loginUser(input: LoginInput): Promise<AuthResponseData> {
  const user = await User.findOne({ email: input.email }).select('+passwordHash');

  if (!user) {
    throw new ApiError(HTTP_STATUS.UNAUTHORIZED, 'Email or password is incorrect');
  }

  if (!user.isActive) {
    throw new ApiError(HTTP_STATUS.FORBIDDEN, 'Account is inactive');
  }

  const isMatched = await bcrypt.compare(input.password, user.passwordHash);

  if (!isMatched) {
    throw new ApiError(HTTP_STATUS.UNAUTHORIZED, 'Email or password is incorrect');
  }

  return buildAuthResponse(user._id.toString());
}

export async function logoutUser(input: RefreshTokenInput): Promise<void> {
  const user = await User.findOne({ refreshToken: input.refreshToken }).select('+refreshToken');

  if (!user) {
    return;
  }

  user.refreshToken = undefined;
  await user.save();
}

export async function getCurrentUser(userId: string) {
  const user = await findUserByIdOrThrow(userId);
  return sanitizeUser(user);
}

export async function refreshUserToken(input: RefreshTokenInput): Promise<AuthResponseData> {
  let payload;

  try {
    payload = verifyRefreshToken(input.refreshToken);
  } catch {
    throw new ApiError(HTTP_STATUS.UNAUTHORIZED, 'Invalid or expired refresh token');
  }

  const user = await User.findById(payload.userId).select('+refreshToken');

  if (!user || !user.refreshToken || user.refreshToken !== input.refreshToken) {
    throw new ApiError(HTTP_STATUS.UNAUTHORIZED, 'Refresh token is not valid');
  }

  if (!user.isActive) {
    throw new ApiError(HTTP_STATUS.FORBIDDEN, 'Account is inactive');
  }

  return buildAuthResponse(user._id.toString());
}

export async function updateCurrentUserProfile(userId: string, input: UpdateProfileInput) {
  const user = await findUserByIdOrThrow(userId);

  if (typeof input.fullName === 'string') {
    user.fullName = input.fullName;
  }

  if (typeof input.avatarUrl === 'string') {
    user.avatarUrl = input.avatarUrl.trim() ? input.avatarUrl.trim() : undefined;
  }

  if (typeof input.bio === 'string') {
    user.bio = input.bio.trim() ? input.bio.trim() : undefined;
  }

  await user.save();
  emitProfileUpdated(mapUserToAuthPayload(user), 'profile');
  return sanitizeUser(user);
}

export async function changeCurrentUserPassword(userId: string, input: ChangePasswordInput) {
  const user = await findUserByIdOrThrow(userId);
  const isMatched = await bcrypt.compare(input.currentPassword, user.passwordHash);

  if (!isMatched) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'Current password is incorrect');
  }

  const isSamePassword = await bcrypt.compare(input.newPassword, user.passwordHash);

  if (isSamePassword) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'New password must be different from current password');
  }

  user.passwordHash = await bcrypt.hash(input.newPassword, SALT_ROUNDS);
  await user.save();
  emitProfileUpdated(mapUserToAuthPayload(user), 'password');
}

export async function requestCurrentUserEmailChange(
  userId: string,
  input: RequestEmailChangeInput,
): Promise<VerificationRequestResult> {
  const user = await findUserByIdOrThrow(userId);

  if (user.email === input.newEmail) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'New email must be different from current email');
  }

  const existing = await User.findOne({ email: input.newEmail, _id: { $ne: userId } });

  if (existing) {
    throw new ApiError(HTTP_STATUS.CONFLICT, 'Email already exists');
  }

  const verificationCode = generateSixDigitCode();
  const expiresAt = new Date(Date.now() + 10 * 60 * 1000);

  user.pendingEmailChange = {
    newEmail: input.newEmail,
    verificationCode,
    expiresAt,
  };

  await user.save();

  return {
    deliveryTarget: input.newEmail,
    expiresAt,
    verificationPreviewCode: verificationCode,
  };
}

export async function confirmCurrentUserEmailChange(
  userId: string,
  input: ConfirmEmailChangeInput,
): Promise<AuthResponseData> {
  const user = await findUserByIdOrThrow(userId);
  const pending = user.pendingEmailChange;

  if (!pending) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'No pending email verification request');
  }

  if (pending.newEmail !== input.newEmail) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'Pending email does not match');
  }

  if (pending.expiresAt.getTime() < Date.now()) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'Email verification code has expired');
  }

  if (pending.verificationCode !== input.verificationCode) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'Email verification code is incorrect');
  }

  const existing = await User.findOne({ email: input.newEmail, _id: { $ne: userId } });

  if (existing) {
    throw new ApiError(HTTP_STATUS.CONFLICT, 'Email already exists');
  }

  user.email = input.newEmail;
  user.pendingEmailChange = undefined;
  await user.save();
  emitProfileUpdated(mapUserToAuthPayload(user), 'email');

  return buildAuthResponse(user._id.toString());
}

export async function requestCurrentUserPhoneChange(
  userId: string,
  input: RequestPhoneChangeInput,
): Promise<VerificationRequestResult> {
  const user = await findUserByIdOrThrow(userId);

  if (user.phone === input.newPhone) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'New phone must be different from current phone');
  }

  const otpCode = generateSixDigitCode();
  const expiresAt = new Date(Date.now() + 10 * 60 * 1000);

  user.pendingPhoneChange = {
    newPhone: input.newPhone,
    otpCode,
    expiresAt,
  };

  await user.save();

  return {
    deliveryTarget: input.newPhone,
    expiresAt,
    verificationPreviewCode: otpCode,
  };
}

export async function confirmCurrentUserPhoneChange(userId: string, input: ConfirmPhoneChangeInput) {
  const user = await findUserByIdOrThrow(userId);
  const pending = user.pendingPhoneChange;

  if (!pending) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'No pending phone verification request');
  }

  if (pending.newPhone !== input.newPhone) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'Pending phone does not match');
  }

  if (pending.expiresAt.getTime() < Date.now()) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'Phone OTP has expired');
  }

  if (pending.otpCode !== input.otpCode) {
    throw new ApiError(HTTP_STATUS.BAD_REQUEST, 'Phone OTP is incorrect');
  }

  user.phone = input.newPhone;
  user.pendingPhoneChange = undefined;
  await user.save();
  emitProfileUpdated(mapUserToAuthPayload(user), 'phone');

  return sanitizeUser(user);
}
